### PR TITLE
Document Rust runtime services traits [ECR-3275]

### PR DESCRIPTION
**Pull request by alekseysidorov**

### CODE REVIEW

File: exonum/src/runtime/rust/service.rs [hunk 1]
```diff
@@ -43,7 +47,18 @@ pub trait ServiceDispatcher: Send {
     ) -> Result<(), ExecutionError>;
 }
 
+/// Describes an Exonum service instance.
+///
+/// That is, it determines how the service instance responds to certain specific
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]: certain and specific are synonyms, pick one.

File: exonum/src/runtime/rust/service.rs [hunk 2]
```diff
@@ -43,7 +47,18 @@ pub trait ServiceDispatcher: Send {
...
+/// requests and events from the runtime side.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
/// requests and events from the runtime.
```

File: exonum/src/runtime/rust/service.rs [hunk 3]
```diff
@@ -43,7 +47,18 @@ pub trait ServiceDispatcher: Send {
...
 pub trait Service: ServiceDispatcher + Debug + 'static {
+    /// Initialize a new service instance with the given parameters.
+    ///
+    /// This method is called once after creating a new service instance.
+    ///
+    /// The configuration parameters passed to the method are discarded immediately.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]: As always, I don't really like `discarded` here.

File: exonum/src/runtime/rust/service.rs [hunk 4]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
         Ok(())
     }
 
+    /// Return a list of root hashes of all Merkelized tables defined by this service,
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Returns a list of root hashes of all Merkelized tables defined by the provided instance,
```

File: exonum/src/runtime/rust/service.rs [hunk 5]
```diff
@@ -43,7 +47,18 @@ pub trait ServiceDispatcher: Send {
     ) -> Result<(), ExecutionError>;
 }
 
+/// Describes an Exonum service instance.
+///
+/// That is, it determines how the service instance responds to certain specific
+/// requests and events from the runtime side.
 pub trait Service: ServiceDispatcher + Debug + 'static {
+    /// Initialize a new service instance with the given parameters.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Initializes a new service instance with the given parameters.
```

File: exonum/src/runtime/rust/service.rs [hunk 6]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
         Ok(())
     }
 
+    /// Return a list of root hashes of all Merkelized tables defined by this service,
+    /// as of the given snapshot of the blockchain state and service instance descriptor.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// as of the given snapshot of the blockchain state.
```

File: exonum/src/runtime/rust/service.rs [hunk 7]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
+    /// [See also.][3]
+    ///
+    /// [1]: ../struct.StateHashAggregator.html
+    /// [2]: ../../blockchain/struct.Block.html#structfield.state_hash
+    /// [3]: ../../blockchain/struct.Schema.html#method.state_hash_aggregator
     fn state_hash(&self, instance: InstanceDescriptor, snapshot: &dyn Snapshot) -> Vec<Hash>;
 
+    /// Perform storage operations on behalf of the service before committing the block.
+    ///
+    /// This changes affect `state_hash` that means this changes must be same for each node.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Any changes of storage state will affect `state_hash`, which means this method must act the same on different nodes.
```

File: exonum/src/runtime/rust/service.rs [hunk 8]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
+    /// In other words you should use only data from context.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]: I'd remove, given my previous suggestion applied.

File: exonum/src/runtime/rust/service.rs [hunk 9]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
+    ///
+    /// The order of invoking `before_commit` method for every service instance depends on the ID.
+    /// `before_commit` for the service instance with the smallest ID is invoked first up to the
+    /// largest one.
+    /// Effectively, this means that services should not rely on a particular ordering of
+    /// Service::execute invocations.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Service::before_commit invocations.
```

File: exonum/src/runtime/rust/service.rs [hunk 10]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
+    /// [3]: ../../blockchain/struct.Schema.html#method.state_hash_aggregator
     fn state_hash(&self, instance: InstanceDescriptor, snapshot: &dyn Snapshot) -> Vec<Hash>;
 
+    /// Perform storage operations on behalf of the service before committing the block.
+    ///
+    /// This changes affect `state_hash` that means this changes must be same for each node.
+    /// In other words you should use only data from context.
+    ///
+    /// The order of invoking `before_commit` method for every service instance depends on the ID.
+    /// `before_commit` for the service instance with the smallest ID is invoked first up to the
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]: I'd remove this sentence, as it is implementation detail and may change in future.

File: exonum/src/runtime/rust/service.rs [hunk 11]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
+    /// largest one.
+    /// Effectively, this means that services should not rely on a particular ordering of
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Effectively, this means that services must not rely on a particular ordering of
```

File: exonum/src/runtime/rust/service.rs [hunk 12]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
+    /// listen address of every full node in the blockchain network.
     fn wire_api(&self, _builder: &mut ServiceApiBuilder) {}
-    // TODO: add other hooks such as "on node startup", etc.
+
+    // TODO: add other hooks such as "on node startup", etc. [ECR-3222]
 }
 
+/// Describes service instance factory for the specific Rust artifact.
 pub trait ServiceFactory: Send + Debug + 'static {
+    /// Return the unique artifact identifier corresponding to the factory.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Returns the unique artifact identifier corresponding to the factory.
```

File: exonum/src/runtime/rust/service.rs [hunk 13]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
     fn artifact_id(&self) -> RustArtifactId;
-
+    /// Return the corresponding protobuf specification.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Returns the corresponding protobuf specification.
```

File: exonum/src/runtime/rust/service.rs [hunk 14]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
     fn artifact_protobuf_spec(&self) -> ArtifactProtobufSpec;
-
+    /// Create a new service instance.
```
Comment (vitvakatu) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Creates a new service instance.
```

File: exonum/src/runtime/rust/service.rs (hunk 9 above)
Comment (popzxc) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// `Service::execute` invocations.
```

shouldn't it be `Service::before_commit` btw?

File: exonum/src/runtime/rust/service.rs [hunk 15]
```diff
@@ -53,21 +68,57 @@ pub trait Service: ServiceDispatcher + Debug + 'static {
...
+    /// Handle block commit event.
+    ///
+    /// This handler is an optional callback method which is invoked by the blockchain
+    /// after each block commit. For example, a service can create one or more transactions
+    /// if a specific condition has occurred.
+    ///
+    /// *Try not to perform long operations in this handler*.
     fn after_commit(&self, _context: AfterCommitContext) {}
 
+    /// Extend API by handlers of this service.
```
Comment (popzxc) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Attach the service API request handlers to the Exonum API schema.
```

File: exonum/src/runtime/rust/service.rs (hunk 13 above)
Comment (popzxc) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Return the protobuf specification used by instances of this service.
```

File: exonum/src/runtime/rust/service.rs (hunk 9 above)
Comment (alekseysidorov) [reply exonum/src/runtime/rust/service.rs]: Oh, this is copypaste from old doc ^_^

File: exonum/src/runtime/rust/service.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (alekseysidorov) [reply exonum/src/runtime/rust/service.rs]: The order of invoking `before_commit` method is implementation detail.

Maybe better

File: exonum/src/runtime/rust/service.rs [hunk 16]
```diff
@@ -280,21 +331,21 @@ impl<'a> AfterCommitContext<'a> {
         }
     }
 
-    /// Return a validator ID if the current node is validator.
+    /// Returns a validator ID if the current node is validator.
```
Comment (Maria-Nosyk) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Returns the validator ID if the current node is a validator.
```

File: exonum/src/runtime/rust/service.rs [hunk 17]
```diff
@@ -280,21 +331,21 @@ impl<'a> AfterCommitContext<'a> {
...
 
-    /// Return a current blockchain height. This height is "height of the last committed block".
+    /// Returns a current blockchain height. This height is "height of the last committed block".
     pub fn height(&self) -> Height {
         // TODO Perhaps we should optimize this method [ECR-3222]
         CoreSchema::new(self.snapshot).height()
     }
 
-    /// Sign and broadcast transaction to other nodes in the network.
+    /// Signs and broadcasts transaction to other nodes in the network.
```
Comment (Maria-Nosyk) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// Signs and broadcasts a transaction to the other nodes in the network.
```

File: exonum/src/runtime/rust/service.rs [hunk 18]
```diff
@@ -43,7 +48,17 @@ pub trait ServiceDispatcher: Send {
...
+///
+/// That is, `Service` determines how a service instance responds to certain requests and events
+/// from the runtime.
 pub trait Service: ServiceDispatcher + Debug + 'static {
+    /// Initializes a new service instance with the given parameters.
+    ///
+    /// This method is called once after creating a new service instance.
+    ///
+    /// The parameters passed to the method are not saved by the framework
+    /// automatically, hence the user must take care of it if needed.
```
Comment (Maria-Nosyk) [inline exonum/src/runtime/rust/service.rs]:
```suggestion
    /// automatically, hence the user must do it manually, if needed.
```

**Pull request merged**